src/nupic/ntypes/Sdr.hpp

```diff
@@ -519,61 +491,124 @@ class SparseDistributedRepresentation : public Serializable
     }
 
     /**
-     * Calculates the number of true / non-zero values in the SDR.  If the SDRs
-     * value is unset this raises an exception.
+     * Calculates the number of true / non-zero values in the SDR.
      *
      * @returns The number of true values in the SDR.
      */
     UInt getSum()
-        { return getFlatIndex().size(); };
+        { return getFlatSparse().size(); };
 
     /**
      * Calculates the sparsity of the SDR, which is the fraction of bits which
      * are true out of the total number of bits in the SDR.
      * I.E.  sparsity = sdr.getSum() / sdr.size
      *
-     * If the SDRs value is unset this raises an exception.
-     *
      * @returns The fraction of values in the SDR which are true.
      */
     Real getSparsity()
         { return (Real) getSum() / size; };
 
     /**
-     * TODO ...
+     * Calculates the number of true bits which both SDRs have in common.
      *
-     * @returns ...
+     * @param sdr, An SDR to compare with, both SDRs must have the same
+     * dimensons.
+     *
+     * @returns Integer, the number of true values which both SDRs have in
+     * common.
      */
     UInt overlap(SparseDistributedRepresentation &sdr) {
-        NTA_ASSERT( false /* Unimplemented */ );
-        return 0;
+        NTA_ASSERT( dimensions.size() == sdr.dimensions.size() );
+        for( UInt i = 0; i < dimensions.size(); i++ )
+            NTA_ASSERT( dimensions[i] == sdr.dimensions[i] );
+
+        UInt ovlp = 0;
+        auto a = this->getDense();
+        auto b = sdr.getDense();
+        for( UInt i = 0; i < size; i++ )
+            ovlp += a[i] && b[i];
+        return ovlp;
     };
 
     /**
      * Make a random SDR, overwriting the current value of the SDR.  The
-     * resulting has uniformly random activations.
+     * result has uniformly random activations.
      *
      * @param sparsity The sparsity of the randomly generated SDR.
+     *
+     * @param rng The random number generator to draw from, optional.
      */
     void randomize(Real sparsity) {
-        NTA_ASSERT( false /* Unimplemented */ );
+        Random rng;
+        randomize( sparsity, rng );
+    };
+
+    void randomize(Real sparsity, Random &rng) {
+        NTA_ASSERT( sparsity >= 0. and sparsity <= 1. );
+        const bool invert = sparsity > .5;
+        if( invert )
+            sparsity = 1 - sparsity;
```

Comment:
Do we want to hack this? If I wanted 90% "sparse", allow me to do so. Or just NTA_ASSERT(sparsity <= 0.5)

Reply:
This does two inversions: first it flips the sparsity, and then it flips the entire SDR.  

This is a performance improvement, because of the way it's implemented.  This turns on random bits, and if the bit it selects is already on it tries again.  If the sparsity is low then there should not be too many retries, but if the sparsity is 100% then it will need many retries as it fills up the entire SDR by randomly selecting every bit in it.

Reply:
Consider this implementation: 

```
vector<UInt> range = vector(size);
iota(range.begin(), range.end(), 0); //<numeric> 
sparse.assign(nbits,0); 
Random.sample(range.begin(), size, sparse.begin(), nbits);
setSparseInPlace(); 
```
It should be faster as it avoids the loops and works (will even more in c++17 std::sample) in vectorized manner